logstash-plugins/logstash-integration-kafka
# Updates Kafka client from 2.8.1 to 3.3.1

## Release notes
Update kafka client to 3.3.1, requires Logstash >= 8.3.0. Deprecated `default` value for setting `client_dns_lookup`, when used it's forced to `use_all_dns_ips`

## What does this PR do?


Updated the Kafka client library used from `2.8.1` to `3.3.1` and Schema Registry from `6.2.2` to `7.3.0`.

This PR updates the client library and, due to the removal of `zookeeper` flag in launching integration Kafka instance, update also the bash scripts used in integration testing.
The update of the Schema Registry forces the dependency on Logstash 8.3 because it's the first that ships Jackson `2.13.3` which is strictly required during the instantiation of the client.


## Why is it important/What is the impact to the user?

Updating the client requires a library (Jackson) that's available from Logstash `8.3.0` so this is a requirement starting from this version of the plugin.
The new Kafka client dropped the `default` value for configuration `client_dns_lookup`. Starting from Kafka `2.6.0` it had been  deprecated and [was removed in Kafka `3.0`](https://cwiki.apache.org/confluence/display/KAFKA/KIP-602%3A+Change+default+value+for+client.dns.lookup).
The default value for plugin setting `client_dns_lookup` switched from `default` to `use_all_dns_ips` which if a user doesn't customize i,t doesn't create any problem.
If a user explicitly set `client_dns_lookup` to `default` and updates the plugin to this version then the value is forced to `use_all_dns_ips`.
Simple explanation why `default` is not a good choice, see [here](https://dttung2905.medium.com/kafka-internals-client-dns-lookup-c993955810e3).

## Checklist




## Author's Checklist


- [x] Run it locally, with Kafka and Logstash.

## How to test this PR locally


- Download Kafka and unpack locally
```sh
curl "https://archive.apache.org/dist/kafka/3.3.1/kafka_2.12-3.3.1.tgz" -o kafka_2.12-3.3.1.tgz
```
- start Kafka and Zookeeper after `cd`-ing into the uncompressed folder
```sh
bin/zookeeper-server-start.sh -daemon config/zookeeper.properties
bin/kafka-server-start.sh -daemon config/server.properties --override advertised.host.name=127.0.0.1
```
- create the partition `logstash_integration_topic_plain` to be later used into the Logstash pipeline
```sh
bin/kafka-topics.sh --create --partitions 3 --replication-factor 1 --topic logstash_topic --bootstrap-server localhost:9092
```
- install this plugin into Logstash, for example in case of local checkout, edit into `Gemfile`
```ruby
gem "logstash-integration-kafka", :path => "/path/to/logstash-integration-kafka"
```
```sh
bin/logstash-plugin install --no-verify
```
- run Logstash pipeline
```
input {
  kafka {
    topics => ["logstash_topic"]
    decorate_events => true
  }
}

output {
  stdout { codec => rubydebug {metadata => true } }
}
```
- Launch CLI Kafka producer client and start typing some lines
```sh
bin/kafka-console-producer.sh --topic logstash_topic --broker-list localhost:9092
```

The events has to be received and logged by Logstash.

Shutdown and cleanup:
```sh
bin/kafka-topics.sh --bootstrap-server localhost:9092 --delete --topic 'logstash_.*'
bin/kafka-server-stop.sh
bin/zookeeper-server-stop.sh
```

## Related issues


- Closes #131

## Use cases

As a Kafka user I want to be able to use a Logstash plugin aligned to the latest major version of Kafka.

## Screenshots



## Logs
